### PR TITLE
fix: make sure we list freeVersions like DEL marker with --versions

### PR DESCRIPTION
**Pull request by harshavardhana**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
fix: make sure we list freeVersions like DEL marker with --versions

## Motivation and Context
freeVersions() was being incorrectly skipped, list it as valid objects properly.

## How to test this PR?
You would have to create a situation where only a FreeVersion() is
left on the namespace, for a specific object.

Ask me for an inspect object for the same, if you want to test
the behavior.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

**Comment by klauspost** [comment]
@krisis to the rescue pls ;)

Isn't the issue that the delete marker *is* deleted, but the delete is not replicated? AFAICT then it shouldn't be returned.

Though I was never sure I understood "free markers" in the first place.

File: .github/workflows/vulncheck.yml [hunk 1]
```diff
@@ -21,7 +21,7 @@ jobs:
       - name: Set up Go
         uses: actions/setup-go@v5
         with:
-          go-version: 1.22.3
+          go-version: 1.22.4
```
Comment (klauspost) [inline .github/workflows/vulncheck.yml]:
```suggestion
          go-version: 1.22.x
```

Is there a point to manually upgrade this?

This isn't linked to the build system anyway. What does that use?

**Comment by harshavardhana** [comment]
> Isn't the issue that the delete marker _is_ deleted, but the delete is not replicated? AFAICT then it shouldn't be returned.

This issue has nothing to do with the DELETE marker @klauspost; it is related to the Tier object when it is deleted. You add a free version marker to track a successful DELETE on the warm tier so that you can delete it once it happens locally on the hot tier.

the free version == DEL marker (but from a tiering point of view)

Delete marker deleted but not replicated is still a delete marker, which must be returned in --versions output.

This is not about regular List(). It's about ListVersions(), which needs to return everything underneath a path. 

It must be consistently the same so that we don't have ghost folders and no way to know what it is.

File: .github/workflows/vulncheck.yml (hunk 1 above)
Comment (harshavardhana) [reply .github/workflows/vulncheck.yml:24]: .x doesn't work @klauspost thats why its kept with actual version.

Comment (harshavardhana) [reply .github/workflows/vulncheck.yml]: Its @aead suggestion that we must always run with actual version for vulncheck this basically tells us that we must make new releases as there are vulnchecks failing.

The version here sticks to the version we use to make releases.

Comment (harshavardhana) [reply .github/workflows/vulncheck.yml]: https://github.com/minio/kes/blob/fe54489812091915144bed6c3a793312fce31e09/.github/workflows/go.yml#L73

Comment (klauspost) [reply .github/workflows/vulncheck.yml]: @harshavardhana I don't suppose this is extracted for the build. 

You could run it on the *build* version when doing a release. This tells you nothing, except being annoying every time a minor release is made.

Comment (harshavardhana) [reply .github/workflows/vulncheck.yml]: We can remove this once we add `vulncheck` to release process.

Comment (harshavardhana) [reply .github/workflows/vulncheck.yml]: // cc @donatello

**Comment by krisis** [comment]
@harshavardhana,
tl;dr free-versions are internal to MinIO and shouldn't be visible to applications via S3 API.

Facts about free-versions:
- Internal to MinIO
- Created (only) when MinIO is unable to delete the tiered object version's content due to temporary unavailability of its warm tier
- Hold a reference to a tiered object version's content in its warm tier

Implementation details:
Free version is implemented as an extension to the Delete Marker type version (`xlMetaV2DeleteMarker`) with additional metadata which differentiate it from a regular DEL marker. 
Free versions are only visible to the scanner process for eventual cleanup of references held by them.

**Comment by harshavardhana** [comment]
I disagree @krisis  this PR is valid, all additional versions in xl.meta that we carry must be listed via --versions otherwise it will confuse the heck out of every other code out there.

**Comment by harshavardhana** [comment]
> Implementation details:
> Free version is implemented as an extension to the Delete Marker type version (`xlMetaV2DeleteMarker`) with additional metadata which differentiate it from a regular DEL marker.
> Free versions are only visible to the scanner process for eventual cleanup of references held by them.

DEL marker is the visual representation for the end-user; that is precisely why we don't list this in the regular List(). It is only listed via ListVersions(). 

Visual representation is as essential for supportability. A free version left over is precisely like a DEL marker with no other non-current versions. 

The final behavior is precisely how a DEL marker is visible.

File: cmd/xl-storage-format-utils.go [hunk 2]
```diff
@@ -30,15 +30,12 @@ func getFileInfoVersions(xlMetaBuf []byte, volume, path string, allParts bool) (
...
 	n := 0
 	for _, fi := range fivs.Versions {
-		// Filter our tier object delete marker
-		if !fi.TierFreeVersion() {
-			fivs.Versions[n] = fi
-			n++
-		} else {
+		if fi.TierFreeVersion() {
 			fivs.FreeVersions = append(fivs.FreeVersions, fi)
+		} else {
```
Comment (krisis) [inline cmd/xl-storage-format-utils.go:50]: `applyVersionActions` (from scanner) expects `getFileInfoVersions` to return `fivs FileInfoVersions` such that, `fivs.Versions` contains only non-free versions and `fivs.FreeVersions` contains only free versions. 
This is used when applying `NewerNoncurrentVersions` lifecycle element.

To address this, we can extend `getFileInfoVersions` to take an argument, `inclFreeVersions bool`, based on which we can populate `fivs.Versions` to include free versions (or not).
```

**Comment by harshavardhana** [comment]
PTAL @krisis @vadmeste @poornas we need to merge this.

**Review by krisis — APPROVED** [review]
LGTM. I have changed my views on making free versions visible to the user. While it exposes an implementation detail it does provide visibility into a tiered object version's lifecycle, especially when its expired/deleted while its warm-tier was temporarily offline.
